app/models/ppm_estimator.rb

```diff
@@ -0,0 +1,216 @@
+    estimator_params[:married] == 'yes' && entitlement.pro_gear_weight_spouse ? entitlement.pro_gear_weight_spouse : 0
+  end
+
+  def weight_self_limited
+    [estimator_params[:weight].to_i, entitlement_self].min
+  end
+
+  def weight_progear_limited
+    [estimator_params[:weight_progear].to_i, entitlement_progear].min
+  end
+
+  def weight_progear_spouse_limited
+    [estimator_params[:weight_progear_spouse].to_i, entitlement_progear_spouse].min
+  end
+
+  def full_pack_cost
+    @full_pack_cost ||= FullPack.rate(effective_400ng_date, orig_svc_area.services_schedule, full_weight) * cwt
+  end
+
+  def orig_svc_area
+    @orig_svc_area ||= ServiceArea.from_date_and_svc_area_number(effective_400ng_date, start_zip3.service_area)
+  end
+
+  def dest_svc_area
+    @dest_svc_area ||= ServiceArea.from_date_and_svc_area_number(effective_400ng_date, end_zip3.service_area)
+  end
+
+  def distance
+    @distance ||= DtodZip3Distance.dist_mi(start_zip3.zip3, end_zip3.zip3).to_i
+  end
+
+  def effective_400ng_date
+    # If the user enters a date that's before or after all 400NG rates in the database, then
+    # use the 400NG rates with an effective date closest to the date the user gave
+    # the FullUnpacks table is the shortest, so should be the quickest to query
+    @effective_400ng_date ||=
+      if (effective_range = FullUnpack.effective_date_range).cover?(date)
+        date
+      elsif date < effective_range.min
+        effective_range.min
+      else
+        effective_range.max
+      end
+  end
+
+  def effective_tdl_date
+    # If the user enters a date that's before or after the TDLs in the database, then
+    # use the discounts with a TDL closest to the date the user gave
+    @effective_tdl_date ||=
+      if (tdl_range = TopTspByChannelLinehaulDiscount.tdl_date_range).cover?(date)
+        date
+      elsif date < tdl_range.min
+        tdl_range.min
+      else
+        tdl_range.max
+      end
+  end
+
+  def rate_area(zip3, zip5)
+    # you can usually get the rate_area from just the ZIP3
+    area = zip3.rate_area
+    # Sometimes, all 5 digits of the ZIP code are needed to get the rate area
+    return area unless area == 'ZIP'
+    Zip5RateArea.select(:rate_area).find_by(zip5: zip5).rate_area
+  end
+
+  def inv_linehaul_discount
+    @inv_linehaul_discount ||= TopTspByChannelLinehaulDiscount.inv_discount(
+      # the origin is a rate area
+      rate_area(start_zip3, estimator_params[:start].to_i),
+      # For CONUS moves, destination is the region.
+      # If the orig and dest are in the same state, the region is 15!
+      start_zip3.state == end_zip3.state ? 15 : end_zip3.region,
+      effective_tdl_date
+    )
+  end
+
+  def base_rate
+    return BaseLinehaul.rate(effective_400ng_date, distance, full_weight) unless full_weight < 1000
+    # pro-rate the 1000 lb baseline rate for shipments less than 1000 lbs
+    BaseLinehaul.rate(effective_400ng_date, distance, 1000) * (full_weight / 1000.0)
```

Comment:
Should these sorts of calculations (pro-rating for < 1000 lbs.) be handled by `BaseLinehaul.rate`? That would simplify this method, presuming it makes sense to move the logic.

Reply:
Done.